In [ ]:
%load_ext sql

In [ ]:
%sql postgresql://postgres:0317@localhost:5432/gameandgrade

# Games and Academic Success

## Introduction


It is very common for people to believe that playing video games leads to poor grades in school. This idea makes sense, considering the amount of time spent on gaming could instead be used for studying. Some people become so addicted to gaming that they stay up late playing, disrupting their sleep patterns. Since I was young, I have always been curious whether this was true, and I still am. In fact, one of the first things I did to succeed in university was to quit playing games. We will explore this dataset using SQL and Tableau to determine whether my choice was valid, and if so, to what extent it was justified.

## Dataset Overview

This dataset, which was obtained from **[here](https://www.kaggle.com/datasets/deadier/play-games-and-success-in-students)**, includes demographic information, students' gaming habits, and their grades. It consists of 10 columns and 770 rows.

In [7]:
# Show column names and data types

In [8]:
%%sql
SELECT
    column_name, data_type
FROM
    information_schema.columns
WHERE
    table_name = 'game_grade';

 * postgresql://postgres:***@localhost:5432/gameandgrade
10 rows affected.


[('Sex', 'bigint'),
 ('School Code', 'bigint'),
 ('Playing Years', 'bigint'),
 ('Playing Often', 'bigint'),
 ('Playing Hours', 'bigint'),
 ('Playing Games', 'bigint'),
 ('Parent Revenue', 'bigint'),
 ('Father Education', 'bigint'),
 ('Mother Education', 'bigint'),
 ('Grade', 'text')]

In [9]:
# Show the first 5 rows

In [10]:
%%sql
SELECT
    *
FROM
    game_grade
LIMIT 5;

 * postgresql://postgres:***@localhost:5432/gameandgrade
5 rows affected.


[(0, 1, 1, 2, 1, 1, 4, 4, 5, '77.50'),
 (1, 1, 1, 3, 1, 1, 1, 3, 3, '83.00'),
 (0, 1, 0, 0, 0, 0, 1, 3, 3, '80.00'),
 (0, 1, 3, 5, 1, 1, 2, 2, 3, '45.00'),
 (1, 1, 1, 1, 2, 1, 1, 3, 4, '85.00')]

### Limitations

Scale interpretation issue:  
The dataset contains categorical variables such as *Playing often, Parent Revenue, Father Education,* and *Mother Education*, represented on different scales (e.g., 0 to 4, 0 to 5, and 0 to 6). However, the original data source does not provide a clear explanation of what each level represents. This lack of documentation makes it difficult to precisely interpret the meaning of each category, which could impact the accuracy of analysis involving these variables.

## Data Cleaning and Preprocessing

### Create a copy of the table to preserve the raw data and rename the columns to snake case

In [15]:
%%sql
DROP TABLE IF EXISTS
    game_grade_clean;

CREATE TABLE
    game_grade_clean AS
SELECT
    *
FROM
    game_grade;

 * postgresql://postgres:***@localhost:5432/gameandgrade
Done.
770 rows affected.


[]

In [16]:
%%sql
ALTER TABLE game_grade_clean RENAME COLUMN "Sex" TO sex;
ALTER TABLE game_grade_clean RENAME COLUMN "School Code" TO school_code;
ALTER TABLE game_grade_clean RENAME COLUMN "Playing Years" TO playing_years;
ALTER TABLE game_grade_clean RENAME COLUMN "Playing Often" TO playing_frequency;
ALTER TABLE game_grade_clean RENAME COLUMN "Playing Hours" TO playing_hours;
ALTER TABLE game_grade_clean RENAME COLUMN "Playing Games" TO playing_games;
ALTER TABLE game_grade_clean RENAME COLUMN "Parent Revenue" TO family_income;
ALTER TABLE game_grade_clean RENAME COLUMN "Father Education" TO father_education;
ALTER TABLE game_grade_clean RENAME COLUMN "Mother Education" TO mother_education;
ALTER TABLE game_grade_clean RENAME COLUMN "Grade" TO grade;

 * postgresql://postgres:***@localhost:5432/gameandgrade
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.
Done.


[]

In [17]:
# Check new column names

In [18]:
%%sql
SELECT
    column_name, data_type
FROM
    information_schema.columns
WHERE
    table_name = 'game_grade_clean';

 * postgresql://postgres:***@localhost:5432/gameandgrade
10 rows affected.


[('sex', 'bigint'),
 ('school_code', 'bigint'),
 ('playing_years', 'bigint'),
 ('playing_frequency', 'bigint'),
 ('playing_hours', 'bigint'),
 ('playing_games', 'bigint'),
 ('family_income', 'bigint'),
 ('father_education', 'bigint'),
 ('mother_education', 'bigint'),
 ('grade', 'text')]

### Handle inconsistent or invalid data

While reviewing the data summary on the source website, we noticed that the column *playing_games* contains a value of 2 in some rows, even though it should only have values of 0 or 1 (0 if the person doesn’t play games and 1 if they do). Since we have little to no information on what the correct value should be, we have decided to remove these rows.

In [21]:
# Identify the invalid rows

In [22]:
%%sql
SELECT
    *
FROM
    game_grade_clean
WHERE
    playing_games = 2;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[(1, 8, 2, 5, 4, 2, 1, 3, 2, '40.00')]

In [23]:
# Remove the invalid rows

In [24]:
%%sql
DELETE FROM
    game_grade_clean
WHERE
    playing_games = 2;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[]

Check that all values are either 0 or 1 and see if there are any null values in the *playing_games* column.

In [26]:
# Count the number of occurrences of each unique value in the 'playing_games' column

In [27]:
%%sql
SELECT
    playing_games, COUNT(*)
FROM
    game_grade_clean
GROUP BY
    playing_games
ORDER BY
    playing_games;

 * postgresql://postgres:***@localhost:5432/gameandgrade
2 rows affected.


[(0, 227), (1, 542)]

We have checked that only 0 and 1 exist in the column and no null values.

Check if other columns have any invalid data including null values and remove them if identified

In [30]:
# Count the number of occurrences of each unique value in the 'sex' column

In [31]:
%%sql
SELECT
    *
FROM
    game_grade_clean
WHERE
    sex NOT IN (0, 1) OR sex IS NULL;

 * postgresql://postgres:***@localhost:5432/gameandgrade
0 rows affected.


[]

In [32]:
# Count the number of occurrences of each unique value in the 'school_code' column

In [33]:
%%sql
SELECT
    school_code, COUNT(*)
FROM
    game_grade_clean
GROUP BY
    school_code
ORDER BY
    school_code;

 * postgresql://postgres:***@localhost:5432/gameandgrade
11 rows affected.


[(1, 112),
 (2, 106),
 (3, 66),
 (4, 148),
 (5, 34),
 (6, 21),
 (7, 41),
 (8, 139),
 (9, 50),
 (10, 32),
 (11, 20)]

In [34]:
# Count the number of occurrences of each unique value in the 'playing_years' column

In [35]:
%%sql
SELECT
    playing_years, count(*)
FROM
    game_grade_clean
GROUP BY
    playing_years
ORDER BY
    playing_years;

 * postgresql://postgres:***@localhost:5432/gameandgrade
5 rows affected.


[(0, 227), (1, 193), (2, 136), (3, 99), (4, 114)]

In [36]:
# Count the number of occurrences of each unique value in the 'playing_frequency' column

In [37]:
%%sql
SELECT
    playing_frequency, count(*)
FROM
    game_grade_clean
GROUP BY
    playing_frequency
ORDER BY
    playing_frequency;

 * postgresql://postgres:***@localhost:5432/gameandgrade
6 rows affected.


[(0, 227), (1, 115), (2, 72), (3, 112), (4, 88), (5, 155)]

In [38]:
# Count the number of occurrences of each unique value in the 'playing_hours' column

In [39]:
%%sql
SELECT
    playing_hours, count(*)
FROM
    game_grade_clean
GROUP BY
    playing_hours
ORDER BY
    playing_hours;

 * postgresql://postgres:***@localhost:5432/gameandgrade
6 rows affected.


[(0, 227), (1, 193), (2, 181), (3, 113), (4, 27), (5, 28)]

In [40]:
# Count the number of occurrences of each unique value in the 'family_income' column

In [41]:
%%sql
SELECT
    family_income, count(*)
FROM
    game_grade_clean
GROUP BY
    family_income
ORDER BY
    family_income;

 * postgresql://postgres:***@localhost:5432/gameandgrade
5 rows affected.


[(0, 91), (1, 184), (2, 302), (3, 142), (4, 50)]

In [42]:
# Count the number of occurrences of each unique value in the 'father_education' column

In [43]:
%%sql
SELECT
    father_education, count(*)
FROM
    game_grade_clean
GROUP BY
    father_education
ORDER BY
    father_education;

 * postgresql://postgres:***@localhost:5432/gameandgrade
7 rows affected.


[(0, 4), (1, 6), (2, 123), (3, 177), (4, 262), (5, 153), (6, 44)]

In [44]:
# Count the number of occurrences of each unique value in the 'mother_education' column

In [45]:
%%sql
SELECT
    mother_education, count(*)
FROM
    game_grade_clean
GROUP BY
    mother_education
ORDER BY
    mother_education;

 * postgresql://postgres:***@localhost:5432/gameandgrade
7 rows affected.


[(0, 1), (1, 27), (2, 164), (3, 199), (4, 245), (5, 106), (6, 27)]

Finally, the last column to check is *grade*. Although this column appears to contain numerical values representing students' scores, its data type is actually text, so it needs to be converted into a numeric format. To ensure a successful conversion, we must first verify that all values in the 'grade' column are in a proper numerical format. We will use a regular expression to identify any non-numeric text values, filtering out those that do not follow the correct numerical pattern (integers or decimals). Additionally, we will check for any null values, as they could also affect the success of the conversion.

In [47]:
# Identify rows that do not follow the correct numerical pattern in the 'grade' column

In [48]:
%%sql
SELECT
    grade
FROM
    game_grade_clean
WHERE
    grade !~ '^[0-9]+(\.[0-9]+)?$' OR grade IS NULL;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[('92..00',)]

There is a row with the value 92..00, which is not in a numeric format due to the extra period. Since this seems to be a typo for 92.00, I will correct it to 92.00 instead of deleting the row. We can also confirm that there aren't any null values in the *grade* column.

In [50]:
# Correct the value into a proper numeric format

In [51]:
%%sql
UPDATE
    game_grade_clean
SET
    grade = '92.00'
WHERE
    grade = '92..00';

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[]

Convert the *grade* column to numeric format before checking for other invalid data.

In [53]:
# Convert the 'grade' column to a numeric data type

In [54]:
%%sql
ALTER TABLE
    game_grade_clean
ALTER COLUMN
    grade TYPE NUMERIC USING grade::NUMERIC;

 * postgresql://postgres:***@localhost:5432/gameandgrade
Done.


[]

Check if there are any values outside the range of 0 and 100.

In [56]:
# Identify rows outside the range of 0 and 100

In [57]:
%%sql
SELECT
    grade
FROM
    game_grade_clean
WHERE
    grade NOT BETWEEN 0 AND 100;

 * postgresql://postgres:***@localhost:5432/gameandgrade
0 rows affected.


[]

Every value in *grade* falls between 0 and 100.

It wouldn’t make sense for someone who doesn’t play games to have non-zero values for *playing_years, playing_frequency,* or *playing_hours,* so check if such data exists.

In [60]:
# Identify rows where 'playing_games' is 0 but at least one of 'playing_years', 'playing_frequency', or 'playing_hours' is not 0

In [61]:
%%sql
SELECT
    *
FROM
    game_grade_clean
WHERE
    playing_games = 0 AND (playing_years != 0 OR playing_frequency != 0 OR playing_hours != 0);

 * postgresql://postgres:***@localhost:5432/gameandgrade
0 rows affected.


[]

Confirmed that no such row exists.

### Remove duplicates

In [64]:
# Find duplicates (rows with rownum > 1 have at least 1 identical row)

In [65]:
%%sql
SELECT
    *
FROM (
    SELECT *, 
        ROW_NUMBER() OVER(PARTITION BY sex, school_code, playing_years, playing_frequency, playing_hours, playing_games,
                          family_income, father_education, mother_education, grade) AS rownum
    FROM
        game_grade_clean
)
WHERE
    rownum > 1;

 * postgresql://postgres:***@localhost:5432/gameandgrade
16 rows affected.


[(0, 4, 0, 0, 0, 0, 1, 3, 2, Decimal('70.00'), 2),
 (0, 4, 0, 0, 0, 0, 2, 3, 4, Decimal('70.00'), 2),
 (0, 5, 0, 0, 0, 0, 2, 5, 4, Decimal('100'), 2),
 (0, 7, 0, 0, 0, 0, 0, 4, 4, Decimal('100'), 2),
 (0, 7, 0, 0, 0, 0, 0, 5, 5, Decimal('100'), 2),
 (0, 7, 0, 0, 0, 0, 0, 6, 6, Decimal('100'), 2),
 (0, 8, 0, 0, 0, 0, 2, 3, 3, Decimal('75.00'), 2),
 (0, 9, 0, 0, 0, 0, 0, 4, 4, Decimal('100'), 2),
 (0, 9, 0, 0, 0, 0, 0, 4, 4, Decimal('100'), 3),
 (0, 9, 0, 0, 0, 0, 0, 4, 4, Decimal('100'), 4),
 (0, 9, 0, 0, 0, 0, 0, 5, 5, Decimal('100'), 2),
 (0, 9, 0, 0, 0, 0, 0, 5, 5, Decimal('100'), 3),
 (0, 9, 0, 0, 0, 0, 0, 5, 5, Decimal('100'), 4),
 (0, 9, 0, 0, 0, 0, 0, 5, 6, Decimal('100'), 2),
 (0, 9, 0, 0, 0, 0, 0, 6, 6, Decimal('100'), 2),
 (0, 9, 0, 0, 0, 0, 0, 6, 6, Decimal('100'), 3)]

We identified 16 rows that are not unique. However, we cannot be certain that they are true duplicates, as this dataset does not include a primary key for each record. This means that identical rows could actually represent distinct entries that just happen to have the same values across all columns.
Additionally, we observed that all of these rows have a value of 0 for *playing_years, playing_frequency, playing_hours,* and *playing_games*. Since a value of 0 in *playing_games* indicates that the person does not play games, it makes sense that the other game-related columns also contain 0's. This effectively reduces four separate columns into a single categorical indicator (as these columns are very highly correlated for non-gamers), significantly increasing the likelihood of identical rows appearing across different samples. Therefore, we decided to keep these rows.

### Data Transformation

In [68]:
# Convert numerical values in the 'sex' and 'playing_games' columns into meaningful texts.

In [69]:
%%sql
ALTER TABLE
    game_grade_clean
ALTER COLUMN sex TYPE TEXT USING sex::TEXT,
ALTER COLUMN playing_games TYPE TEXT USING playing_games::TEXT;

UPDATE
    game_grade_clean
SET
    sex = CASE
        WHEN sex = '0' THEN 'female'
        WHEN sex = '1' THEN 'male'
        ELSE sex
    END,
    playing_games = CASE
        WHEN playing_games = '0' THEN 'non_gamer'
        WHEN playing_games = '1' THEN 'gamer'
        ELSE playing_games
    END;

 * postgresql://postgres:***@localhost:5432/gameandgrade
Done.
769 rows affected.


[]

In [70]:
# Show top 5 rows

In [71]:
%%sql
SELECT
    *
FROM
    game_grade_clean
LIMIT 5;

 * postgresql://postgres:***@localhost:5432/gameandgrade
5 rows affected.


[('female', 5, 1, 5, 2, 'gamer', 3, 4, 3, Decimal('100')),
 ('male', 5, 4, 2, 1, 'gamer', 2, 4, 4, Decimal('100')),
 ('female', 5, 0, 0, 0, 'non_gamer', 0, 0, 0, Decimal('100')),
 ('male', 1, 0, 0, 0, 'non_gamer', 3, 5, 4, Decimal('60.00')),
 ('male', 1, 0, 0, 0, 'non_gamer', 2, 3, 2, Decimal('64.75'))]

To enhance the clarity of the dataset, we transformed the values in the *sex* column, replacing 0 with 'female' and 1 with 'male'. Similarly, in the *playing_games* column, we recoded 0 as 'non_gamer' and 1 as 'gamer'. This transformation improves the interpretability of the data and ensures that the analysis is more intuitive.

## Exploratory Data Analysis (EDA)

In [74]:
# Overview of student grades

In [75]:
%%sql
SELECT
    ROUND(AVG(grade), 2) AS avg_grade,
    MIN(grade) AS min_grade,
    MAX(grade) AS max_grade
FROM
    game_grade_clean;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[(Decimal('77.57'), Decimal('33.00'), Decimal('100'))]

The average grade, minimum grade, and maximum grade are 77.57, 33.00, and 100, respectively.

In [77]:
# Number of students who play games vs those who don't

In [78]:
%%sql
SELECT
    playing_games, COUNT(*)
FROM
    game_grade_clean
GROUP BY
    playing_games;

 * postgresql://postgres:***@localhost:5432/gameandgrade
2 rows affected.


[('non_gamer', 227), ('gamer', 542)]

There are 227 students who don't play games and 542 who do.

In [80]:
# Number of female vs male

In [81]:
%%sql
SELECT
    sex, COUNT(*)
FROM
    game_grade_clean
GROUP BY
    sex;

 * postgresql://postgres:***@localhost:5432/gameandgrade
2 rows affected.


[('female', 386), ('male', 383)]

There are 386 females and 383 males.

In [83]:
# Average grade between gamers and non-gamers

In [84]:
%%sql
SELECT
    playing_games,
    ROUND(AVG(grade), 2) AS avg_grade
FROM
    game_grade_clean
GROUP BY
    playing_games
ORDER BY 1;

 * postgresql://postgres:***@localhost:5432/gameandgrade
2 rows affected.


[('gamer', Decimal('75.92')), ('non_gamer', Decimal('81.51'))]

The average grades for non-gamers and gamers are 81.51 and 75.92, respectively, with a difference of 5.59.

In [86]:
# Average grade between two genders

In [87]:
%%sql
SELECT
    sex,
    ROUND(AVG(grade), 2) AS avg_grade
FROM
    game_grade_clean
GROUP BY
    sex;

 * postgresql://postgres:***@localhost:5432/gameandgrade
2 rows affected.


[('female', Decimal('79.74')), ('male', Decimal('75.38'))]

The average grades for female and male students are 79.74 and 75.38, respectively, with a difference of 4.36.

In [89]:
# Proportion of gamers in each gender.

In [90]:
%%sql
SELECT
    sex,
    playing_games,
    COUNT(*) AS count,
    ROUND(COUNT(*) * 100 / SUM(COUNT(*)) OVER(PARTITION BY sex), 2) AS percentage
FROM
    game_grade_clean
GROUP BY
    sex, playing_games
ORDER BY
    sex, playing_games;

 * postgresql://postgres:***@localhost:5432/gameandgrade
4 rows affected.


[('female', 'gamer', 212, Decimal('54.92')),
 ('female', 'non_gamer', 174, Decimal('45.08')),
 ('male', 'gamer', 330, Decimal('86.16')),
 ('male', 'non_gamer', 53, Decimal('13.84'))]

The percentage of gamers among female students and male students is 54.92% and 86.16%, respectively. This significant difference in gaming participation between the two groups may have contributed to the observed disparity in the average grades between genders.

To minimize the influence of gender on grade, we compare the grade differences between gamers and non-gamers within the same gender.

In [92]:
# Average grades for gamers and non-gamers, broken down by gender

In [93]:
%%sql
SELECT
    sex,
    playing_games,
    ROUND(AVG(grade), 2) AS avg_grade
FROM
    game_grade_clean
GROUP BY
    sex, playing_games
ORDER BY
    sex, playing_games;

 * postgresql://postgres:***@localhost:5432/gameandgrade
4 rows affected.


[('female', 'gamer', Decimal('77.17')),
 ('female', 'non_gamer', Decimal('82.87')),
 ('male', 'gamer', Decimal('75.11')),
 ('male', 'non_gamer', Decimal('77.03'))]

Non-gamers have higher average grades than gamers in both genders. However, the difference is more noticeable in females (5.70) compared to males (1.92).

We want to calculate the correlation between *playing_games* and *grade*, as well as between *sex* and *grade*.

In [96]:
# Before calculating correlation coefficients, sex and playing_games should be converted to numerical values 0 and 1

In [97]:
%%sql
WITH
    game_grade_num AS (
        SELECT
            CASE WHEN sex = 'female' THEN '0' ELSE '1' END AS sex_num, school_code, playing_years, playing_frequency, playing_hours,
            CASE WHEN playing_games = 'non_gamer' THEN '0' ELSE '1' END AS playing_games_num, family_income, father_education, mother_education, grade
        FROM game_grade_clean)

/* sex: converted female to 0, male to 1
    playing_games: converted non_gamer to 0 and gamer to 1 */

-- calculate correlation

SELECT
    CORR(CAST(playing_games_num AS NUMERIC), grade) AS corr_game_grade,
    CORR(CAST(sex_num AS NUMERIC), grade) AS corr_sex_grade
FROM
    game_grade_num;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[(-0.15507895772625185, -0.13247200127612166)]

Both *playing_games* and *sex* show negative correlation coefficients with *grade*. There is a very weak tendency for students who play games to have slightly lower grades, but the relationship is not strong enough to draw significant conclusions. Similarly, male students tend to have slightly lower grades than female students, but the relationship is too weak.

In [99]:
# Correlation between grade and game-related variables

In [100]:
%%sql
SELECT 
    CORR(playing_years, grade) AS corr_years_grade,
    CORR(playing_frequency, grade) AS corr_freq_grade,
    CORR(playing_hours, grade) AS corr_hours_grade
FROM
    game_grade_clean;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[(-0.106972540641275, -0.004469387260529937, -0.1540154170812488)]

All game-related variables show negative correlations with grades, but none of them are strongly correlated. Among the three, *playing_hours* has the highest correlation, so let's explore its trend further.

In [102]:
# Playing hours vs average grade

In [103]:
%%sql
SELECT
    playing_hours,
    ROUND(AVG(grade), 2) AS avg_grade
FROM
    game_grade_clean
GROUP BY
    playing_hours
ORDER BY
    playing_hours;

 * postgresql://postgres:***@localhost:5432/gameandgrade
6 rows affected.


[(0, Decimal('81.51')),
 (1, Decimal('77.44')),
 (2, Decimal('75.23')),
 (3, Decimal('75.75')),
 (4, Decimal('74.02')),
 (5, Decimal('72.34'))]

It's clear that there is a downward trend in the average grade as playing time increase.

Let's explore whether family background influence students' grades.

In [105]:
# Family income vs average grade

In [106]:
%%sql
SELECT
    family_income,
    ROUND(AVG(grade), 2) AS avg_grade
FROM
    game_grade_clean
GROUP BY
    family_income
ORDER BY
    family_income;

 * postgresql://postgres:***@localhost:5432/gameandgrade
5 rows affected.


[(0, Decimal('95.97')),
 (1, Decimal('68.97')),
 (2, Decimal('75.30')),
 (3, Decimal('78.90')),
 (4, Decimal('85.68'))]

There is an upward trend in average grades as family income increases. However, students with a family income score of 0 have a significantly higher average grade compared to other groups. We will take a closer look at this group to explore potential factors behind this trend.

In [108]:
# Rows where family income score equals 0

In [109]:
%%sql
SELECT
    *
FROM
    game_grade_clean
WHERE
    family_income = 0;

 * postgresql://postgres:***@localhost:5432/gameandgrade
91 rows affected.


[('female', 5, 0, 0, 0, 'non_gamer', 0, 0, 0, Decimal('100')),
 ('female', 9, 2, 5, 2, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('95.00')),
 ('male', 7, 1, 2, 1, 'gamer', 0, 5, 3, Decimal('100')),
 ('female', 7, 3, 2, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 7, 3, 2, 2, 'gamer', 0, 4, 3, Decimal('85.00')),
 ('male', 7, 2, 5, 1, 'gamer', 0, 4, 5, Decimal('90.00')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 4, Decimal('90.00')),
 ('male', 7, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 1, 5, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 1, 5, 1, 'gamer', 0, 4, 4, Decimal('86.00')),
 ('female', 7, 3, 2, 1, 'gamer', 0, 5, 5, Decimal('90.00')),
 ('male', 7, 3, 5, 3, 'gamer', 0, 3, 3, Decimal('90.00')),
 ('male', 7, 3, 5, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 3, 4, 3, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 7, 1, 2, 2, 'gamer', 0, 4, 4, Decimal('95.00')),
 ('male', 7, 2, 2, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 3, 2, 3, 'gamer', 0, 3, 3, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 4, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 7, 2, 5, 3, 'gamer', 0, 4, 5, Decimal('100')),
 ('male', 7, 1, 4, 1, 'gamer', 0, 6, 6, Decimal('100')),
 ('male', 7, 2, 3, 2, 'gamer', 0, 5, 4, Decimal('90.00')),
 ('male', 7, 2, 4, 3, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 2, 3, 3, 'gamer', 0, 4, 2, Decimal('85.00')),
 ('male', 7, 1, 4, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('100')),
 ('male', 7, 1, 5, 1, 'gamer', 0, 3, 2, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 6, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 0, 0, 0, 'non_gamer', 0, 4, 3, Decimal('95.00')),
 ('male', 7, 4, 5, 3, 'gamer', 0, 4, 4, Decimal('100')),
 ('male', 7, 1, 3, 1, 'gamer', 0, 2, 2, Decimal('100')),
 ('female', 7, 4, 5, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 0, 0, 0, 'non_gamer', 0, 4, 5, Decimal('98.00')),
 ('male', 7, 1, 5, 1, 'gamer', 0, 4, 4, Decimal('97.00')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 1, 4, 2, 'gamer', 0, 4, 4, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('90.00')),
 ('male', 7, 4, 5, 3, 'gamer', 0, 4, 4, Decimal('97.00')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 3, Decimal('97.00')),
 ('male', 9, 0, 0, 0, 'non_gamer', 0, 4, 3, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('80.00')),
 ('male', 9, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 9, 1, 5, 1, 'gamer', 0, 4, 3, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 3, Decimal('85.00')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 6, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('male', 9, 4, 5, 3, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 9, 2, 4, 2, 'gamer', 0, 5, 5, Decimal('90.00')),
 ('female', 9, 2, 4, 2, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 4, 3, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 3, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 6, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('female', 9, 2, 5, 2, 'gamer', 0, 6, 6, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('85.00')),
 ('male', 9, 3, 5, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 9, 4, 4, 3, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 9, 2, 5, 1, 'gamer', 0, 5, 4, Decimal('80.00')),


We see a lot of rows with 7 or 9 for *school_code*, so let's count them to find out their proportion.

In [111]:
# Number of students with family income score of 0 in each school

In [112]:
%%sql
SELECT
    school_code, COUNT(*)
FROM
    game_grade_clean
WHERE
    family_income = 0
GROUP BY
    school_code
ORDER BY
    school_code;

 * postgresql://postgres:***@localhost:5432/gameandgrade
3 rows affected.


[(5, 1), (7, 41), (9, 49)]

Out of 91 students with a family income score of 0, all but one are from schools with school codes 7 or 9. One possible hypothesis is that a family income score of 0 is assigned to students whose income information is missing. Since the data source does not provide a clear explanation, it is unclear whether a score of 0 represents the lowest income group or a null group. To better understand this, we decided to further investigate any abnormal trends in these schools.

In [114]:
# Rows where school code equals to 7 or 9

In [115]:
%%sql
SELECT
    *
FROM
    game_grade_clean
WHERE
    school_code IN (7, 9)
ORDER BY
    school_code;

 * postgresql://postgres:***@localhost:5432/gameandgrade
91 rows affected.


[('male', 7, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 1, 5, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 1, 5, 1, 'gamer', 0, 4, 4, Decimal('86.00')),
 ('female', 7, 3, 2, 1, 'gamer', 0, 5, 5, Decimal('90.00')),
 ('male', 7, 3, 5, 3, 'gamer', 0, 3, 3, Decimal('90.00')),
 ('male', 7, 3, 5, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 3, 4, 3, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 7, 1, 2, 2, 'gamer', 0, 4, 4, Decimal('95.00')),
 ('male', 7, 2, 2, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 3, 2, 3, 'gamer', 0, 3, 3, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 4, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 7, 2, 5, 3, 'gamer', 0, 4, 5, Decimal('100')),
 ('male', 7, 1, 4, 1, 'gamer', 0, 6, 6, Decimal('100')),
 ('male', 7, 2, 3, 2, 'gamer', 0, 5, 4, Decimal('90.00')),
 ('male', 7, 2, 4, 3, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 2, 3, 3, 'gamer', 0, 4, 2, Decimal('85.00')),
 ('male', 7, 1, 4, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('100')),
 ('male', 7, 1, 5, 1, 'gamer', 0, 3, 2, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 6, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 0, 0, 0, 'non_gamer', 0, 4, 3, Decimal('95.00')),
 ('male', 7, 4, 5, 3, 'gamer', 0, 4, 4, Decimal('100')),
 ('male', 7, 1, 3, 1, 'gamer', 0, 2, 2, Decimal('100')),
 ('female', 7, 4, 5, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 0, 0, 0, 'non_gamer', 0, 4, 5, Decimal('98.00')),
 ('male', 7, 1, 5, 1, 'gamer', 0, 4, 4, Decimal('97.00')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('male', 7, 1, 4, 2, 'gamer', 0, 4, 4, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('90.00')),
 ('male', 7, 4, 5, 3, 'gamer', 0, 4, 4, Decimal('97.00')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 3, Decimal('97.00')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('95.00')),
 ('male', 7, 1, 2, 1, 'gamer', 0, 5, 3, Decimal('100')),
 ('female', 7, 3, 2, 1, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 7, 3, 2, 2, 'gamer', 0, 4, 3, Decimal('85.00')),
 ('male', 7, 2, 5, 1, 'gamer', 0, 4, 5, Decimal('90.00')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('100')),
 ('female', 7, 0, 0, 0, 'non_gamer', 0, 5, 4, Decimal('90.00')),
 ('male', 9, 3, 5, 3, 'gamer', 0, 5, 5, Decimal('85.00')),
 ('male', 9, 3, 1, 2, 'gamer', 0, 5, 5, Decimal('70.00')),
 ('male', 9, 1, 1, 4, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('100')),
 ('male', 9, 2, 3, 2, 'gamer', 0, 6, 6, Decimal('98.50')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('100')),
 ('male', 9, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('male', 9, 4, 1, 2, 'gamer', 0, 5, 4, Decimal('100')),
 ('male', 9, 2, 5, 1, 'gamer', 2, 5, 5, Decimal('100')),
 ('male', 9, 4, 5, 5, 'gamer', 0, 5, 6, Decimal('100')),
 ('female', 9, 2, 5, 2, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 9, 0, 0, 0, 'non_gamer', 0, 4, 3, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 4, 4, Decimal('80.00')),
 ('male', 9, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 9, 1, 5, 1, 'gamer', 0, 4, 3, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 6, 6, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 3, Decimal('85.00')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 6, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 5, Decimal('100')),
 ('male', 9, 4, 5, 3, 'gamer', 0, 5, 5, Decimal('100')),
 ('male', 9, 2, 4, 2, 'gamer', 0, 5, 5, Decimal('90.00')),
 ('female', 9, 2, 4, 2, 'gamer', 0, 5, 5, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 4, 3, Decimal('100')),
 ('female', 9, 0, 0, 0, 'non_gamer', 0, 5, 3, Decimal('100')),
 ('female', 9, 0, 0,

We see lots of 0's in *family_income* as expected, but the other columns seem fine. Let's find out how many non-zero rows there are for *family_income*.

In [117]:
# Find rows with family_income not equal to 0

In [118]:
%%sql
SELECT
    *
FROM
    game_grade_clean
WHERE
    school_code IN (7, 9) AND family_income != 0
ORDER BY
    school_code;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[('male', 9, 2, 5, 1, 'gamer', 2, 5, 5, Decimal('100'))]

There is one row where *family_income* is not 0, which contradicts the hypothesis that family income information is missing for all students in these schools. However, it seems unlikely that only one student would have a family income score of 2, while all others have a score of 0 across both schools. This suggests that the entry could be a data recording error, and that these schools likely didn't have information on students' family income. In this case, the *family_income* column may not accurately represent the actual family income levels of the students in these schools.

In [120]:
# Average grade across all schools

In [121]:
%%sql
SELECT
    ROUND(AVG(grade), 2) AS avg_grade
FROM
    game_grade_clean;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[(Decimal('77.57'),)]

In [122]:
# Average grade for school code 7 and 9

In [123]:
%%sql
SELECT
    school_code,
    ROUND(AVG(grade), 2) AS avg_grade
FROM
    game_grade_clean
WHERE
    school_code IN (7, 9)
GROUP BY
    school_code
ORDER BY
    school_code;

 * postgresql://postgres:***@localhost:5432/gameandgrade
2 rows affected.


[(7, Decimal('96.83')), (9, Decimal('95.27'))]

The average grades for schools 7 and 9 are 96.83 and 95.27, respectively, which are about 20 points higher than the overall average across all schools. This is expected, as these two schools heavily represent the 0 family income score group, which has significantly higher average grades. While we don't have enough information to determine the exact reason for these high grades, the fact that there is only one row with a non-zero family income score in these schools, and that the *family_income* column may not accurately reflect the actual family income levels of the students, suggests that it is reasonable to set the 0 family income score group aside and focus on comparing the non-zero family income score groups. Based on this, we conclude that there is an upward trend between family income and grades.  
Let's calculate the correlation between family_income and grade for all schools, as well as for schools excluding 7 and 9.

In [125]:
# Correlation between family income and grade

In [126]:
%%sql
SELECT
    CORR(family_income, grade) AS overall_corr,
    (SELECT
         CORR(family_income, grade)
     FROM
        game_grade_clean
    WHERE
        school_code NOT IN (7, 9)) AS filtered_corr
FROM
    game_grade_clean;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[(-0.056963354919060974, 0.2855508361283341)]

The correlation coefficient increased from -0.0570 to 0.2856 after excluding schools 7 and 9 from the calculation.  

Next, we will explore the effect of parents' level of education on students' grades using *father_education* and *mother_education*. Before diving deeper, we noticed that these two variables share a similar structure with *family_income*, as they are all demographic variables measured on scales starting from 0. If that's the case, the 0's in these two variables may indicate that the information for those entries was unavailable.  
Let's compare the correlation between grade and parents' level of education (*father_education* and *mother_education*) with and without including cases where the education level is 0.

In [128]:
# Correlation between parents education and grade including 0 education level

In [129]:
%%sql
SELECT    
    CORR(father_education, grade) AS corr_father_grade,
    CORR(mother_education, grade) AS corr_mother_grade
FROM
    game_grade_clean;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[(0.319597157688003, 0.3507485259879546)]

In [130]:
# Correlation between parents education and grade without 0 education level

In [131]:
%%sql
SELECT    
    CORR(father_education, grade) AS corr_father_grade,
    CORR(mother_education, grade) AS corr_mother_grade
FROM
    game_grade_clean
WHERE
    father_education != 0 OR mother_education != 0;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[(0.32778455438158477, 0.35830458209001487)]

In [132]:
# Number of rows where father_education or mother_education is 0

In [133]:
%%sql
SELECT
    COUNT(*)
FROM 
    game_grade_clean
WHERE
    father_education = 0 OR mother_education = 0;

 * postgresql://postgres:***@localhost:5432/gameandgrade
1 rows affected.


[(4,)]

Excluding the rows with 0 in either of the columns doesn't seem to have a significant impact on the correlation coefficients, likely because the number of such rows is too small to affect the results.

In [135]:
# Average grade for each father's education level and the number of students in each group.

In [136]:
%%sql
SELECT
    father_education,
    ROUND(AVG(grade), 2) AS avg_grade,
    COUNT(*) AS num_students
FROM
    game_grade_clean
GROUP BY
    father_education
ORDER BY
    father_education;

 * postgresql://postgres:***@localhost:5432/gameandgrade
7 rows affected.


[(0, Decimal('95.00'), 4),
 (1, Decimal('76.08'), 6),
 (2, Decimal('70.82'), 123),
 (3, Decimal('71.64'), 177),
 (4, Decimal('78.09'), 262),
 (5, Decimal('84.65'), 153),
 (6, Decimal('91.16'), 44)]

The number of students with father_education equal to 0 or 1 is too small (4 and 6, respectively) to accurately reflect the impact of father_education on grades. However, as the father's education level increases, we observe a clear upward trend in the average grade in groups with a sufficient number of students.  
Let's see if this is also true for *mother_education*.

In [138]:
# Average grade for each mother's education level and the number of students in each group.

In [139]:
%%sql
SELECT
    mother_education,
    ROUND(AVG(grade), 2) AS avg_grade,
    COUNT(*) AS num_students
FROM
    game_grade_clean
GROUP BY
    mother_education
ORDER BY
    mother_education;

 * postgresql://postgres:***@localhost:5432/gameandgrade
7 rows affected.


[(0, Decimal('100.00'), 1),
 (1, Decimal('63.10'), 27),
 (2, Decimal('73.10'), 164),
 (3, Decimal('73.55'), 199),
 (4, Decimal('79.61'), 245),
 (5, Decimal('86.35'), 106),
 (6, Decimal('94.94'), 27)]

There is a clear upward trend, except when *mother_education* is 0, where the group is represented by only one student.  
Both *father_education* and *mother_education* show the highest average grades when equal to 0, but the number of students in these groups is too small. Since we suspect that the 0's in these columns may represent missing information, we believe it is reasonable to conclude that the higher the parents' education level, the higher the grades students achieve.

Lastly, Let's analyze the differences in gaming behavior based on family income.

In [142]:
# Number and percentage of gamers for each family income group

In [143]:
%%sql
SELECT
    family_income,
    playing_games,
    COUNT(*),
    ROUND(COUNT(*) * 100 / SUM(COUNT(*)) OVER(PARTITION BY family_income), 2) AS percentage
FROM
    game_grade_clean
WHERE
    family_income > 0 
GROUP BY 1, 2
ORDER BY 1, 2;

 * postgresql://postgres:***@localhost:5432/gameandgrade
8 rows affected.


[(1, 'gamer', 128, Decimal('69.57')),
 (1, 'non_gamer', 56, Decimal('30.43')),
 (2, 'gamer', 220, Decimal('72.85')),
 (2, 'non_gamer', 82, Decimal('27.15')),
 (3, 'gamer', 101, Decimal('71.13')),
 (3, 'non_gamer', 41, Decimal('28.87')),
 (4, 'gamer', 43, Decimal('86.00')),
 (4, 'non_gamer', 7, Decimal('14.00'))]

The percentage of gamers in family income groups 1, 2, and 3 remains around 70%. However, it suddenly jumps to 86% in group 4. While there are only 7 non-gamers in group 4, the total number of students, including gamers, is 50, which appears to be a reasonably large sample size.

## Key Findings

### Relationship Between Gaming and Academic Performance
* Non-gamers achieve higher grades than gamers on average.
    * The average grade for non-gamers is 81.51, while gamers score 75.92, showing a 5.59-point gap.
* This trend is more pronounced among female students.
    * Female non-gamers score 5.70 points higher than female gamers, whereas the gap is smaller for males (1.92 points).
* Longer gaming hours are associated with lower academic performance
### Gender Differences in Gaming Habits and Academic Performance
* Male students are significantly more likely to be gamers than female students.
    * 86.16% of male students play games, compared to 54.92% of female students.
* Female students perform better academically than male students
    * The average grade for females is 79.74, while for males, it is 75.38, indicating a 4.36-point gap.
### Family Background and Academic Success
* Higher family income is associated with better academic performance.
    * Students from wealthier families tend to have higher grades, suggesting that financial stability may contribute to academic success.
* Parental education levels also show a positive correlation with student grades.
    * Students with highly educated parents achieve higher grades on average.
* Interestingly, students from the highest-income families have the highest percentage of gamers.
    * This could indicate that wealthier students have more access to gaming resources or different gaming habits compared to lower-income students.

## Conclusion

Our analysis reveals clear trends in the relationship between gaming habits and academic performance. Non-gamers consistently achieve higher grades than gamers, with a more significant gap observed among female students. Additionally, longer gaming hours are associated with lower academic performance, suggesting that excessive gaming may have a negative impact on student success.  
Gender differences in gaming behavior are also evident. Male students are much more likely to play games than female students, yet the academic performance gap between gamers and non-gamers is smaller for males. This indicates that while gaming may influence grades, other factors such as study habits, time management, and social influences could also play a role.  
Family background appears to be another important factor in academic performance. Students from wealthier families and those with highly educated parents tend to have higher grades. Interestingly, students from the highest-income families also have the highest percentage of gamers, raising questions about how socio-economic status influences gaming habits and its effects on education.  
While this analysis provides valuable insights, it does not establish a causal relationship between gaming and academic success. Further research, incorporating additional factors such as study hours, sleep hours, and personal motivation, would be necessary to understand the complex interactions between gaming and student achievement in more depth.